### PR TITLE
golang: expose add{Decoded,Encoded}Data

### LINKED ISSUE #36911: golang: expose add{Decoded,Encoded}Data so that people can add data when processing headers/data/trailers
*Title*: *golang: expose add{Decoded,Encoded}Data so that people can add data when processing headers/data/trailers*

*Description*:
>Describe the desired behavior, what scenario it enables and how it
would be used.

We would like to provide a way to add extra data when processing headers/data/trailers. For example, turn a headers only request into a request with a body, add more data when processing trailers, and so on.

The `add{Decoded,Encoded}Data` serves our purpose well so we want to expose this method in Go, like:

```
f.callbacks.DecoderFilterCallbacks().AddData([]byte("blah"), true)
# equal to `addDecodedData(Instance "blah", true)
f.callbacks.EncoderFilterCallbacks().AddData([]byte("blah"), true)
# equal to `addEncodedData(Instance "blah", true)
```

Here I expose all the arguments of `add{Decoded,Encoded}Data` as only the callers know what arguments should be passed inside.

CC @doujiang24 

[optional *Relevant Links*:]
>Any extra documentation required to understand the issue.

### PR DESCRIPTION
**Pull request by spacewander**
Commit Message: golang: expose add{Decoded,Encoded}Data
Additional Description:
Risk Level: Low
Testing: Integration test
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
Fixes #36911
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: contrib/golang/filters/http/source/golang_filter.cc [hunk 1]
```diff
@@ -527,6 +527,47 @@ CAPIStatus Filter::continueStatus(ProcessorState& state, GolangStatus status) {
...
+
+  if (state.isThreadSafe()) {
+    addDataInternal(state, data, is_streaming);
+    return CAPIStatus::CAPIOK;
+  }
+
+  auto weak_ptr = weak_from_this();
+  state.getDispatcher().post([this, weak_ptr, &state, data, is_streaming] {
+    if (!weak_ptr.expired() && !hasDestroyed()) {
+      addDataInternal(state, data, is_streaming);
```
Comment (doujiang24) [inline contrib/golang/filters/http/source/golang_filter.cc:557]: It's unsafe to use data pointer from Golang side, after returning to Golang.

Should copy for later using, i.e.
https://github.com/envoyproxy/envoy/blob/96b345ff93d6268165603cf3c49d0027c553c41a/contrib/golang/filters/http/source/golang_filter.cc#L681-L683

**Review by doujiang24 — COMMENTED** [review]
otherwise, lgtm.

File: contrib/golang/filters/http/source/golang_filter.cc (hunk 1 above)
Comment (spacewander) [reply contrib/golang/filters/http/source/golang_filter.cc]: @doujiang24 
Good catch! Fixed in new commit.

**Review by doujiang24 — APPROVED** [review]
lgtm, it's ready to merge, thanks cc @KBaichoo

**Pull request merged**